rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -0,0 +1,1424 @@
+        self._set_optimizer(tf.keras.optimizers.Adam(config[LEARNING_RATE]))
+        self._create_metrics()
+        self._update_metrics_to_log()
+
+        self.all_labels_embed = None  # needed for efficient prediction
+
+    def _check_data(self) -> None:
+        if "text_features" not in self.data_signature:
+            raise ValueError(
+                f"No text features specified. "
+                f"Cannot train '{self.__class__.__name__}' model."
+            )
+        if self.config[INTENT_CLASSIFICATION]:
+            if "label_features" not in self.data_signature:
+                raise ValueError(
+                    f"No label features specified. "
+                    f"Cannot train '{self.__class__.__name__}' model."
+                )
+            if (
+                self.config[SHARE_HIDDEN_LAYERS]
+                and self.data_signature["text_features"]
+                != self.data_signature["label_features"]
+            ):
+                raise ValueError(
+                    "If hidden layer weights are shared, data signatures "
+                    "for text_features and label_features must coincide."
+                )
+
+        if self.config[ENTITY_RECOGNITION] and "tag_ids" not in self.data_signature:
+            raise ValueError(
+                f"No tag ids present. "
+                f"Cannot train '{self.__class__.__name__}' model."
+            )
+
+    def _create_metrics(self) -> None:
+        # self.metrics preserve order
+        # output losses first
+        self.mask_loss = tf.keras.metrics.Mean(name="m_loss")
+        self.intent_loss = tf.keras.metrics.Mean(name="i_loss")
+        self.entity_loss = tf.keras.metrics.Mean(name="e_loss")
+        # output accuracies second
+        self.mask_acc = tf.keras.metrics.Mean(name="m_acc")
+        self.response_acc = tf.keras.metrics.Mean(name="i_acc")
+        self.entity_f1 = tf.keras.metrics.Mean(name="e_f1")
+
+    def _update_metrics_to_log(self) -> None:
+        if self.config[MASKED_LM]:
+            self.metrics_to_log += ["m_loss", "m_acc"]
+        if self.config[INTENT_CLASSIFICATION]:
+            self.metrics_to_log += ["i_loss", "i_acc"]
+        if self.config[ENTITY_RECOGNITION]:
+            self.metrics_to_log += ["e_loss", "e_f1"]
+
+    def _prepare_layers(self) -> None:
+        self.text_name = TEXT
+        self._prepare_sequence_layers(self.text_name)
+        if self.config[MASKED_LM]:
+            self._prepare_mask_lm_layers(self.text_name)
+        if self.config[INTENT_CLASSIFICATION]:
+            self.label_name = TEXT if self.config[SHARE_HIDDEN_LAYERS] else LABEL
+            self._prepare_input_layers(self.label_name)
+            self._prepare_label_classification_layers()
+        if self.config[ENTITY_RECOGNITION]:
+            self._prepare_entity_recognition_layers()
+
+    def _prepare_sparse_dense_layers(
+        self,
+        feature_signatures: List[FeatureSignature],
+        name: Text,
```

Comment:
is this the `layer_name` ?

Reply:
it's the name to which part of nn (text or label) these layers correspond to. Not sure what other name give here

Reply:
so this is basically an enum because it's either `text` or `label`? How do you call them in the paper? `nn_part`?

Reply:
I don't understand what you mean

Reply:
I guess it's fine then